mm2src/coins/z_coin/z_rpc.rs

```diff
@@ -1,149 +1,497 @@
-use bigdecimal::BigDecimal;
-use common::jsonrpc_client::{JsonRpcClient, JsonRpcRequest};
-use common::mm_number::MmNumber;
+use super::{z_coin_errors::*, ZcoinConsensusParams};
+use crate::utxo::utxo_common;
+use common::executor::Timer;
+use common::log::{debug, error, info};
+use common::{async_blocking, spawn_abortable, AbortOnDropHandle};
+use db_common::sqlite::rusqlite::{params, Connection, Error as SqliteError, NO_PARAMS};
+use db_common::sqlite::{query_single_row, run_optimization_pragmas};
+use futures::channel::mpsc::{channel, Receiver as AsyncReceiver, Sender as AsyncSender};
+use futures::channel::oneshot::{channel as oneshot_channel, Sender as OneshotSender};
+use futures::lock::{Mutex as AsyncMutex, MutexGuard as AsyncMutexGuard};
+use futures::StreamExt;
+use http::Uri;
 use mm2_err_handle::prelude::*;
-use rpc::v1::types::{Bytes as BytesJson, H256 as H256Json, H264 as H264Json};
-use serde_json::{self as json, Value as Json};
+use parking_lot::Mutex;
+use prost::Message;
+use protobuf::Message as ProtobufMessage;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use tokio::task::block_in_place;
+use tonic::transport::{Channel, ClientTlsConfig};
+use zcash_client_backend::data_api::chain::{scan_cached_blocks, validate_chain};
+use zcash_client_backend::data_api::error::Error as ChainError;
+use zcash_client_backend::data_api::{BlockSource, WalletRead, WalletWrite};
+use zcash_client_backend::proto::compact_formats::CompactBlock;
+use zcash_client_sqlite::error::SqliteClientError as ZcashClientError;
+use zcash_client_sqlite::wallet::init::{init_accounts_table, init_wallet_db};
+use zcash_client_sqlite::WalletDb;
+use zcash_primitives::consensus::BlockHeight;
+use zcash_primitives::transaction::TxId;
+use zcash_primitives::zip32::ExtendedFullViewingKey;
 
-#[derive(Debug, Serialize)]
-pub struct ZSendManyItem {
-    pub amount: BigDecimal,
-    #[serde(rename = "opreturn")]
-    #[serde(skip_serializing_if = "Option::is_none")]
-    pub op_return: Option<BytesJson>,
-    pub address: String,
+mod z_coin_grpc {
+    tonic::include_proto!("cash.z.wallet.sdk.rpc");
+}
+use z_coin_grpc::compact_tx_streamer_client::CompactTxStreamerClient;
+use z_coin_grpc::{BlockId, BlockRange, ChainSpec, TxFilter};
+
+pub type WalletDbShared = Arc<Mutex<WalletDb<ZcoinConsensusParams>>>;
+
+struct CompactBlockRow {
+    height: BlockHeight,
+    data: Vec<u8>,
 }
 
-#[derive(Debug, Deserialize)]
-pub struct ZOperationTxid {
-    pub txid: H256Json,
+/// A wrapper for the SQLite connection to the block cache database.
+pub struct BlockDb(Connection);
+
+impl BlockDb {
+    /// Opens a connection to the wallet database stored at the specified path.
+    pub fn for_path<P: AsRef<Path>>(path: P) -> Result<Self, db_common::sqlite::rusqlite::Error> {
+        let conn = Connection::open(path)?;
+        run_optimization_pragmas(&conn)?;
+        conn.execute(
+            "CREATE TABLE IF NOT EXISTS compactblocks (
```

Comment:
***A suggestion(Not a blocker, probably better to do in sepereated PR)***

I think we should avoid doing db operations with hard coded queries everywhere. Reason is might cause some problem by just missing or adding an argument and actually it's hard to catch those kind of problems in PR reviews.

An example(can be much better, this is just a proof of concept, took 5min to write.):
```rs
enum SqlOperations {
    CreateTableIfNotExists
}

struct SqlBuilder {
    table_name: String,
    operation: SqlOperations,
    args: Vec<String>
}

impl SqlBuilder {
    fn new(table_name: String, operation: SqlOperations) -> Self {
        SqlBuilder {
            table_name, operation, args: Vec::default()
        }
    }

    fn add_arg(&mut self, arg: String) {
        self.args.push(arg);
    }

    fn as_raw(&mut self) -> String {
        match self.operation {
            SqlOperations::CreateTableIfNotExists => {
                let mut sql = format!("CREATE TABLE IF NOT EXISTS {}", self.table_name);
                    
                for (pos, e) in self.args.iter().enumerate() {
                    if pos == 0 {
                        sql = format!("{} ( \n", sql);
                    }
                    
                    if pos == self.args.len() - 1 {
                        sql = format!("{} {} \n )", sql, e);
                    } else {
                        sql = format!("{} {}, \n", sql, e);
                    }
                }

                sql
            },
        }
    }
}

fn main() {
    let mut create_table = SqlBuilder::new("example_db".to_string(), SqlOperations::CreateTableIfNotExists);
    create_table.add_arg("height INTEGER PRIMARY KEY".to_string());
    create_table.add_arg("weight INTEGER".to_string());
    create_table.add_arg("data BLOB NOT NULL".to_string());
    let raw_sql = create_table.as_raw();
    println!("{}", raw_sql);
}
```
This way we will provide `single source of truth` which will prevent us from SQL syntax problems, and increase the readability and maintainability.

Reply:
@ozkanonur That's a good idea, could you create an issue please? 🙂

Reply:
I've already implemented [something similar](https://github.com/KomodoPlatform/atomicDEX-API/blob/dev/mm2src/db_common/src/sql_query.rs) in my task 😅